internal/types/compare.go

```diff
@@ -52,17 +53,31 @@ func Compare(docValue, filterValue any) CompareResult {
 		panic("compare: filterValue is nil")
 	}
 
-	switch docValue := docValue.(type) {
+	c := &comparer{
+		compareTypeOrder: compareTypeOrder,
+	}
+
+	return c.compare(docValue, filterValue)
+}
+
+// comparer uses compareTypeOrder to perform type comparison.
+type comparer struct {
```

Comment:
Why a struct is needed? Is it used as a dependency somewhere?

Reply:
This is used for checking if number type is equal or not within nested documents, we need to pass how to compare type order when it is called within recursion. Definitely need to improve documentation

For standard comparison by `Compare(a, b)`, 
a doc `{"v": {"foo": {"bar": [int32(1)]}}}` should equal
a doc `{"v": {"foo": {"bar": [float64(1)]}}}`
For `CompareOrder(a,b,Ascending)`, above should not equal.